src/Init/Data/String/Basic.lean

```diff
@@ -290,73 +290,111 @@ where go (acc : String) (s : String) : List String → String
   | a :: as => go (acc ++ s ++ a) s as
   | []      => acc
 
-/-- Iterator for `String`. That is, a `String` and a position in that string. -/
+/-- Iterator over the characters (`Char`) of a `String`.
+
+Typically created by `s.iter`, where `s` is a `String`. -/
 structure Iterator where
+  /-- The string the iterator is for. -/
   s : String
+  /-- The current position.
+
+  This position is not necessarily legal for the string, for instance if one keeps calling
+  `Iterator.next` when `Iterator.atEnd` is true. If the position is not a legal one, then the
+  current character is `(default : Char)`, similar to `String.get` on an illegal position. -/
   i : Pos
   deriving DecidableEq
 
+/-- The iterator positioned at the first character of a string (if any). -/
 def mkIterator (s : String) : Iterator :=
   ⟨s, 0⟩
 
+@[inherit_doc mkIterator]
 abbrev iter := mkIterator
 
+/-- The notion of size for a string iterator is the number of bytes remaining. -/
 instance : SizeOf String.Iterator where
   sizeOf i := i.1.utf8ByteSize - i.2.byteIdx
 
 theorem Iterator.sizeOf_eq (i : String.Iterator) : sizeOf i = i.1.utf8ByteSize - i.2.byteIdx :=
   rfl
 
 namespace Iterator
+/-- The *full* string the iterator is for. -/
 def toString : Iterator → String
   | ⟨s, _⟩ => s
```

Comment:
I'm puzzled by how we have both this and `Iterator.s`; that is, I think this is the same as
```suggestion
/-- The *full* string the iterator is for. -/
abbrev toString := Iterator.s
```
Should the docstring explain which one is preferred?

Reply:
Same for `forward` and `nextn`